File: C++/030. Substring with Concatenation of All Words.cpp
class Solution {
private:
    bool checkSubstring(string &s,unordered_map<string,int> &hash, int k, int m, int L, int N)
    {
        bool res;
        string word = s.substr(k,m);
        //cout << "word=" << word << "\n";
        if(!hash[word])
        {
            return false;
        }
        if(L+1==N)
        {return true;}
        else
        {
            hash[word]--;
            res = checkSubstring(s,hash,k+m,m,L+1,N);
            hash[word]++;
            return res;
        }
    }
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        vector<int> res={};
        int n = s.length(), size = words.size();
        if(!n||!size){return res;}
        int m = words[0].length();
        unordered_map<string,int> hash;
        for (int j=0;j<size;j++)
        {
            hash[words[j]]+=1;
        }
        for (int i=0;i<=n-m*size;i++)
        {
            if(checkSubstring(s,hash,i,m,0,size))
            {
                res.push_back(i);
            }
        }
        return res;
    }
};

File: python/030. Substring with Concatenation of All Words.py
class Solution(object):
    def findSubstring(self, s, words):
        """
        :type s: str
        :type words: List[str]
        :rtype: List[int]
        """
        if len(words) == 0: return []
        m = len(words[0])
        freq = {}
        for word in words:
            if word not in freq: freq[word] = 0
            freq[word] += 1
        total = len(freq)
        ans = []
        for offset in range(m):
            status = {}
            start = end = offset
            cnt = total
            while end + m <= len(s):
                word = s[end:end+m]
                if word not in freq:
                    status = {}
                    cnt = total
                    start = end = end + m
                else:
                    if word not in status:
                        status[word] = 1
                    else:
                        status[word] += 1
                    if status[word] == freq[word]:
                        cnt -= 1
                        if cnt == 0: 
                            ans.append(start)
                            status[s[start:start+m]] -= 1
                            start += m
                            cnt += 1
                    elif status[word] > freq[word]:
                        while status[word] > freq[word]:
                            prevWord = s[start:start+m]
                            if status[prevWord] == freq[prevWord]: cnt += 1
                            status[prevWord] -= 1
                            start += m
                    else:
                        pass
                    end += m
        return ans